### PR TITLE
Add simple timer and led animations.

### PR DESCRIPTION
**Pull request by scubapanks**

### CODE REVIEW

File: BoxTimer.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#ifndef BoxTimer_h
+#define BoxTimer_h
+#include <EnhancedThread.h>
```
Comment (SciLor) [inline BoxTimer.h:3]: Please remove as its unnecessary.
```suggestion
```

File: BoxTimer.cpp [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        return false;
+    }
+
+    unsigned long currentMillis = millis();
+    if (_endMillis <= currentMillis) {
+        _isRunning = false;
+        return false;
+    }
+
+    return true;
```
Comment (SciLor) [inline BoxTimer.cpp:19]:
```suggestion
    return _isRunning;
```

File: BoxTimer.cpp [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+bool BoxTimer::isRunning() { 
+    if (! _isRunning) {
+        return false;
+    }
+
+    unsigned long currentMillis = millis();
+    if (_endMillis <= currentMillis) {
+        _isRunning = false;
+        return false;
```
Comment (SciLor) [inline BoxTimer.cpp:16]:
```suggestion
```

File: BoxTimer.cpp [hunk 4]
```diff
@@ -0,0 +1,24 @@
+#include "BoxTimer.h"
+
+void BoxTimer::setTimer(unsigned long milliseconds) { 
+    _endMillis = millis() + milliseconds;
+    _isRunning = true;
+}
+
+bool BoxTimer::isRunning() { 
+    if (! _isRunning) {
```
Comment (SciLor) [inline BoxTimer.cpp:9]:
```suggestion
    if (!_isRunning) {
```

File: BoxLEDs.cpp [hunk 5]
```diff
@@ -40,16 +78,141 @@ void BoxLEDs::disableRedLED(bool disabled) {
     }
 }
 
-void BoxLEDs:: setIdleType(IDLE_TYPE idleType) {
-    _idleType = idleType;
+void BoxLEDs::setIdleAnimation(ANIMATION_TYPE animationType,ANIMATION_COLOR animationColor) {
```
Comment (SciLor) [inline BoxLEDs.cpp:81]: Code duplication with setActiveAnimation.
A private helper function could use a pointer to generalize the code

File: BoxLEDs.cpp [hunk 6]
```diff
@@ -12,21 +12,59 @@ void BoxLEDs::begin() {
...
+            }
+        } else { 
+            if (animation->state >= LED_PWM_MIN + animation->step) {
+                animation->state -= animation->step;
+            } else {
+                animation->direction = ANIMATION_DIRECTION::UP;
+                animation->state = LED_PWM_MIN;
+            }
+        }
+        setAll(_transformPulse(animation->state, animation->color));
```
Comment (SciLor) [inline BoxLEDs.cpp:57]: Please ignore LED_PWM_MIN/MAX should only be handled within the setLED methods (as its just a workaround).

```suggestion
    } else if (animation->type == ANIMATION_TYPE::PULSE) {
        if (animation->direction == ANIMATION_DIRECTION::UP) {
            if (animation->state <= animation->step) {
                animation->state = animation->state+animation->step;
            } else {
                animation->direction = ANIMATION_DIRECTION::DOWN;
                animation->state = LED_PWM_MAX;
            }
        } else { 
            if (animation->state >= animation->step) {
                animation->state -= animation->step;
            } else {
                animation->direction = ANIMATION_DIRECTION::UP;
                animation->state = 0;
            }
        }
        setAll(_transformPulse(animation->state, animation->color));
```

File: BoxTimer.cpp (hunk 2 above)
Comment (SciLor) [inline BoxTimer.cpp:19]:
```suggestion
    if (!_isRunning)
        return false;

    unsigned long currentMillis = millis();
    if (_endMillis <= currentMillis)
        _isRunning = false;

    return _isRunning;
```

File: BoxTimer.h (hunk 1 above)
Comment (SciLor) [inline BoxTimer.h:3]: Unnecessary
```suggestion
```

File: BoxLEDs.cpp (hunk 5 above)
Comment (scubapanks) [reply BoxLEDs.cpp:81]: Done in commit 640ca076236cc24a14c615ce5628263df4c6d060
private helper function addAnimation to remove code duplication

File: BoxLEDs.cpp (hunk 6 above)
Comment (scubapanks) [reply BoxLEDs.cpp]: The suggestion changes the logic. A replacement without using LED_PWM_MIN/MAX in commit 2c34e5377b9ad3f4502a7a233d0a77402f6a7c0c.

File: BoxLEDs.cpp [hunk 7]
```diff
@@ -40,18 +40,18 @@ void BoxLEDs::_handleAnimation(ANIMATION* animation) {
         setAll(_wheel(random(255)));
     } else if (animation->type == ANIMATION_TYPE::PULSE) {
         if (animation->direction == ANIMATION_DIRECTION::UP) {
-            if (animation->state <= LED_PWM_MAX - animation->step) {
+            if (animation->state <= 0xFF - animation->step) {
                 animation->state = animation->state+animation->step;
```
Comment (SciLor) [inline BoxLEDs.cpp:44]:
```suggestion
                animation->state += animation->step;
```

Comment (scubapanks) [reply BoxLEDs.cpp:43]: Resolved in 837300d9fd0072863f806b5fe18a7aa8006ec632

File: BoxLEDs.cpp [hunk 8]
```diff
@@ -40,16 +78,141 @@ void BoxLEDs::disableRedLED(bool disabled) {
...
+    
+    default:
+        break;
+    }
+
+    _timer.setTimer(duration);
+    _activeAnimationRunning = true;
+}
+
+BoxLEDs::CRGB BoxLEDs::_transformCRGBToAnimationColor(ANIMATION_COLOR animationColor) {
```
[the remark this answers is not in the document]
Comment (scubapanks) [reply BoxLEDs.cpp]: Here we go: c5455d61225cc4b8a4f31a8b8d930040c6f50bfa

**Pull request merged**